contracts/near/src/lib.rs

```diff
@@ -0,0 +1,489 @@
+//#![deny(warnings)]
+
+use {
+    error::Error,
+    ext::ext_wormhole,
+    near_sdk::{
+        borsh::{
+            self,
+            BorshDeserialize,
+            BorshSerialize,
+        },
+        collections::{
+            UnorderedMap,
+            UnorderedSet,
+        },
+        env,
+        is_promise_success,
+        log,
+        near_bindgen,
+        AccountId,
+        Balance,
+        BorshStorageKey,
+        Gas,
+        PanicOnDefault,
+        Promise,
+        StorageUsage,
+    },
+    p2w_sdk::{
+        BatchPriceAttestation,
+        Identifier,
+        PriceStatus,
+    },
+    state::{
+        PriceFeed,
+        Source,
+        Vaa,
+    },
+    std::io::Cursor,
+};
+
+pub mod error;
+pub mod ext;
+pub mod governance;
+pub mod state;
+
+const _GAS_FOR_VERIFY_VAA: Gas = Gas(10_000_000_000_000);
+const INITIAL_STALENESS_THRESHOLD: u64 = 60;
+
+#[derive(BorshSerialize, BorshStorageKey)]
+enum StorageKeys {
+    Source,
+    Prices,
+}
+
+/// The `State` contains all persisted state for the contract. This includes runtime configuration.
+///
+/// There is no valid Default state for this contract, so we derive PanicOnDefault to force
+/// deployment using one of the #[init] functions in the impl below.
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize, PanicOnDefault)]
+pub struct Pyth {
+    /// The set of `Source`s from which Pyth attestations are allowed to be relayed from.
+    pub sources: UnorderedSet<Source>,
+
+    /// The Governance Source.
+    pub gov_source: Source,
+
+    /// The last executed sequence number for governance actions.
+    pub last_gov_seq: u64,
+
+    /// A Mapping from PriceFeed ID to Price Info.
+    pub prices: UnorderedMap<Identifier, PriceFeed>,
+
+    /// The `AccountId` of the Wormhole contract used to verify VAA's.
+    pub wormhole: AccountId,
+
+    /// A hash of the current contract code.
+    ///
+    /// If this `hash` does not match the current contract code, then it indicates that a pending
+    /// upgrade has been allowed, calling `upgrade` with code that matches this hash will cause the
+    /// contract to upgrade itself.
+    pub codehash: [u8; 32],
+
+    /// Staleness threshold for rejecting price updates.
+    pub stale_threshold: u64,
+
+    /// Fee for updating price.
+    pub update_fee: Balance,
+}
+
+/// The `Action` acts as an enum of possible `Action`'s this contract can receive via VAA.
+#[derive(BorshDeserialize, BorshSerialize)]
```

Comment:
I see `Action` to be used in the tests for serializing inputs into vaa, since it's mostly for test purposes shouldn't it go there?

Also the governance payloads are different.

Reply:
well actually GovernanceAction has most of it.